Fix wrapdateline and other lon=180 fixes

Kirill888 opened the pull request:
### lon=180 handling

- Redo chop_along_antimeridan logic
  1. Use `shapely.ops.split` instead of what was
  2. Be robust when projecting `lon=180` line into coordinate space of geometry
- Redo `to_crs(.., wrapdetaline=True)` case
  1. Clip points close to `lon=+/-180` to `lon=+/-180` after projecting chopped geometries
- Add `lonlat_bounds` method that can deal with `lon=180` crossing geometries


### clean ups
- redo densify logic
- Expose CRS "area of use"
- More convenience methods
- More  mypy cleaning
- `Geometry.clone` and constructor for other `Geometry`

Kirill888 committed:
Expose pyproj.CRS object via `CRS.proj` property

Kirill888 committed:
More mypy fixes: remove more comment based annotations

Also add some missing ones

Kirill888 committed:
Add CRS.valid_region method

libproj keeps track of Lon/Lat bounding box for every CRS, expose that
information to the user via `Geometry` object

uchchwhash commented on datacube/utils/geometry/_base.py:
```diff
@@ -233,6 +233,20 @@ def __eq__(self, other: SomeCRS) -> bool:
     def __ne__(self, other) -> bool:
         return not (self == other)
 
+    @property
```
If we are going to expose this, why not get rid of `_crs`.

Kirill888 replied:
Do you mean rename `_crs`  into `proj`?

This exposes private variable `_crs` in read-only fashion, just having it renamed would allow accidental assignment without updating other state variables. Honestly I would do the same `crs_str`, except I'm not sure if any code depends on that "property".

uchchwhash replied:
Yes, that's what I meant, but your justification is sound. I am with you on `crs_str`, and IMHO we should make it private regardless.

Kirill888 committed:
Convert CRS.crs_str into read-only and deprecated property

Kirill888 committed:
More mypy fixes/annotations in geometry

Kirill888 committed:
Rename BoundingBox.{width|height} -> {span_x|span_y}

width/height is too commonly assumed to be integer in pixels, but BoundingBox is
not about pixels, adding width/height that return span_x/span_y but casted
to an int.

Kirill888 committed:
Move `_dist` function where it is used

It is badly named function, it returns distance squared and not just distance,
moving it out of the way before it gets used somewhere else

Kirill888 committed:
Add geometry.sides method

Kirill888 committed:
Add BoudningBox.from_{xy|points} methods

Kirill888 committed:
Refactor densify/segmented and add more robust clone method

Prior to this change segmented/to_crs didn't work on LinearRing geometries
because of cloning via json.

- Add more robust clone functionality
- Refactor Geometry constructor to accept Self
- Refactor densify/segmented
  - fewer intermediate geometry constructors
  - no sqrt
  - densify operates on coordinates not geometries

Kirill888 committed:
Add geometry.lonlat_bounds method

Kirill888 committed:
Move doc test for _round_to_res into a unit test

Kirill888 committed:
Add BoundingBox.range_{x|y} accessors

Kirill888 committed:
Expose shapely.ops.split method

Kirill888 committed:
Add geometry.{mutligeom|common_crs} helpers

Kirill888 committed:
Completely replace chop along antimeridian code

Idea is still the same, but using more robust implementation using
shapely.ops.split functionality.

1. Project lon=180 line into CRS space (taking care of nan/inf)
2. Split Geometry into MultiGeometry using that line
3. Transform partitioned geometry to Lon/Lat
4. Cleanup points around lon=+/-180 line by clipping them to either 180 or -180,
   based on where other points in that geometry are

Kirill888 committed:
Area of use might not be defined for CRS

uchchwhash commented on datacube/utils/geometry/_base.py:
```diff
@@ -309,24 +382,33 @@ def go(x):
+def densify(coords: CoordList, resolution: float) -> CoordList:
     """
-    Adds points so they are at most `distance` apart.
+    Adds points so they are at most `step` apart.
```
What is `step`?

Kirill888 replied:
thanks I'll fix

alexgleith commented:
Looks good to me.

Pinging @luigidifraia, since Catapult have experience working with data on the antimeridian... do you have any input, Luigi?

uchchwhash reviewed (approved):
That's a lot of thoughts going into this. Awesome! 💯

Kirill888 committed:
Use more robust lonlat_bound when constructing query #886

Kirill888 committed:
Fix documentation comment step->resolution
